pkg/kubelet/lifecycle/handlers.go

```diff
@@ -105,6 +110,32 @@ func resolvePort(portReference intstr.IntOrString, container *v1.Container) (int
 	return -1, fmt.Errorf("couldn't find port: %v in %v", portReference, container)
 }
 
+// formatURL formats a URL from args. It attempts to create a properly-formatted *url.URL
+// from the provided values, while preserving compatibility with legacy behavior.
+func formatURL(scheme string, host string, port int, rawPathAndQuery string) *url.URL {
+	// split rawPathAndQuery into slice with path and query
+	pathAndQuery := strings.SplitN(rawPathAndQuery, "?", 2)
```

Comment:
This could also include a fragment. What about just doing something like this:

```go
parsedPath, err := url.Parse(rawPathAndQuery)
if err != nil { ... }
// Verify that the path parsed currectly
if parsedPath.Scheme != "" || parsedPath.Host != "" || parsedPath.User != nil {
  // Probably just return an error here
}
path := parsedPath.Path
query := parsedPath.RawQuery
fragment := parsedPath.Fragment
```

However, it's

Reply:
IIRC, Using `url.Parse` here breaks some legacy behavior that this explicitly strives to preserve. Using `url.Parse` was how this was originally implemented in this PR, but the current implementation in this PR was required for some necessary tests to pass.